keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1069,187 @@ the remaining requests in that queue start getting faster service.  In
+safe side we should probably implement borrowing across priority levels.
+However, given we don't want to block introducing the `width` concept on
+design and implementation of borrowing, until this is done we have two
+main options:
+- cap the `width` at the concurrency units assigned to the priority level
+- reject requests for which we won't be able to allocate enough concurrency
+  units
+
+To avoid breaking some users, we will proceed with the first option.  That
+said, to accommodate for the inaccuracy here we will introduce a concept
+of `additional latency` for a request.  This basically means that after the
+request finishes in a real world, we still don't mark it as finished in
+the virtual world for `additional latency`.
+Adjusting virtual time of a queue to do that is trivial.  The other thing
+to tweak is to ensure that the concurrency units will not get available
+for other requests for that time (because currently all actions are
+triggerred by starting or finishing some request).  We will maintain that
+possibility by wrapping the handler into another one that will be sleeping
+for `additional latence` after the request is processed.
+
+Note that given the estimation for duration of processing the requests is
+automatically corrected (both up and down), there is no need to change that
+in the initial version.
+
+#### Dispatching the request
+
+The hardest part of adding support for LIST requests is dispatching the
+requests.  Now in order to start processing a request, it has to accumulate
+`<width>` units of concurrency.
+
+The important requirement to recast now is fairness.  As soon a single
+request can consume more units of concurrency, the fairness is
+no longer about the number of requests from a given queue, but rather
+about number of consumed concurrency units.  This justifes the above
+definition of adjusting the cost of the request to now be equal to
+`<width> x <processing latency>` (instead of just `<processing latency>`).
+
+At the same time, we want to maximally utilize the available capacity.
+In other words, we want to minimize the time when some concurrency unit
+is not used, but there are requests at a given PL that could use it.
+
+In order to achieve the above goals, we are introducing the following
+modification to the current dispatching algorith:
+- as soon as we dispatch the first token (concurrency unit) to a queue
+  with LIST request at its front, all further tokens are also assigned
+  to this queue until the LIST request can be started.
+  Such approach allows us to not block too many tokens and avoid deadlocks
+  if we would be dispatching tokens to multiple LIST requests without having
+  enough of them for a given priority level.
+- however, to ensure fairness (especially over longer period of times)
+  the virtual time of the queue is now advanced by the new cost of the
+  request, i.e. by `<width> x <processing latency>` (as opposed to simply
```

Comment:
I don't think these units work out, maaaaybe some amount of width is fungible with some amount of time but just multiplying them doesn't seem right at all?

Reply:
I actually think it is the right way.
Basically, the request now consumes <width> seats instead of 1 seat, so it's cost is actually <width> times more.


> maaaaybe some amount of width is fungible with some amount of time 

That is a bit orthogonal aspect. The fact that the cost of the request is multiplication of those two doesn't mean we can use arbitrary width and adjust latency.

The concept of request cost is used for fairness - we now have to be fair not wrt "request latencies", but wrt to "request cost" (which previously was equal to request latency, and now is multiplied by <wdith> - as the request is now consuming <width> seats for the <request latency> time).

Reply:
What you wrote in the comment sounds fine, and I agree the product makes sense as a prioritization mechanism, but the text has the virtual time being incremented by this amount, and that doesn't seem right at all, the units don't match up, it's like adding meter-seconds to a date.

Reply:
> but the text has the virtual time being incremented by this amount, and that doesn't seem right at all, the units don't match up, it's like adding meter-seconds to a date.

I think I disagree.
The reason is that `<width>` is actually a scalar (and we shouldn't think about it as a length measure).

How I'm conceptually thinking about request with `width=<width>` is that it occupies `<width>` seats, so technically
there are `<width>` smaller requests that together are producing a result of our request.

Reply:
Yes, it is a scalar-- but that doesn't mean it doesn't have a unit! It represents the number of seats a request requires. 

Here's another argument: it doesn't make sense to estimate the virtual finish time as `latency * width` AND require `width` seats to execute-- that's double counting.

Now, I am willing to buy that latency actually does increase with the number of items, and therefore it is legit to increase both. But without evidence I'm not willing to believe that the same multiplier works for both, that's just too improbable. Does the cost of a request really go with the square of the number of items? I don't really believe it unless it's attached to a constant factor that shrinks it a lot.

Reply:
> What we do with virtual_time is that we actually advance it by the actual latency (not expected latency) [to be more specific - we first advance it by expected, but then decrease it by the difference with the actual latency].

Right, this illustrates that we should be adding on an estimate of *time* not an estimate of *work*. It matters if we want to answer the question, "is this queue too full to admit another item?"

Note that width is capped by resources, but latency is capped by the global timeout. This is more evidence that the estimates have to be different.

I think you can make me happy by either making the time estimate a function of the width estimate, or by having it be a function of the number of items. I don't even care if it happens to come out to the same value in the first take, I just think conceptually it's important that it be separate, for clarity about what is going on if no other reason. (However I think if we take into account the different caps, it will probably come out different. I highly doubt that a request of 1 item is the same fraction of CPU/RAM as it is of the 60s global timeout.)

Reply:
Or, another way of putting it-- we should be estimating wall time, not compute time, because wall time is what adds up as we go further back in the queue.

Reply:
> Right, this illustrates that we should be adding on an estimate of time not an estimate of work. It matters if we want to answer the question, "is this queue too full to admit another item?"

I agree with "It matters if we want to answer the question, "is this queue too full to admit another item?"" - but I disagree with the first part.
When processing items in a queue, what matters is amount of work.
For a moment assuming every request takes the same amount of time to process, we may be able to process e.g. 100 requests with `width=1`, but we may not able to process 100 requests with `width=100`.
So in order to answer the question "is this queue too full to admit another item" - we exactly need the cost of requests.

> I think you can make me happy by either making the time estimate a function of the width estimate, or by having it be a function of the number of items.

OK - that I'm fully onboard with (in fact maybe a function of both) - changed. See line 1254-1256

Reply:
> When processing items in a queue, what matters is amount of work.

You could also make me happy by making virtual time explicitly advanced by "work" instead of just wall time. So when you update the estimate after the request has finished, update with `actual seats * actual wall time` instead of just `actual wall time`. I think that is closer to what you're trying to accomplish?

I guess I can see it working fine whether the queues track time or work. I just feel very strongly that you can't have the estimate be one and the correction be the other.

> See line 1254-1256

1239 needs an update too, if you don't want to take my other suggestion instead :)

Reply:
> So when you update the estimate after the request has finished, update with actual seats * actual wall time instead of just actual wall time. I think that is closer to what you're trying to accomplish?

I guess I'm a bit lost now.
I think this is exactly what I was proposing from the beginning (it's just "actual seats" will always be "estimated seats" because we can't really correct it).
If this is what would make you happy - what are the exact sentences that suggest otherwise?

Reply:
The text only mentions the estimate and doesn't explain how the virtual time is changed once the request ends and the duration is known. I think I'd be looking for something like this:

We will change the semantics of "virtual time" tracked by the queues to correspond to work, instead of just wall time. That means when we estimate a request's virtual duration, we will use `estimated width * estimated latency` instead of just `estimated latecy`. And when a request finishes, we will update the virtual time for it with `seats * actual latency` (note that seats will always equal the estimated width, since we have no way to figure out if a request used less concurrency than we granted it).